File: chess/pieces/knight.rb
module Chess::Pieces

  ##
  #
  #  Knight
  #  @author Adrien Thébault <me@adrien-thebault.fr>
  #

  module Knight

    def self.possible_moves(game,pos)

      color, ennemy = game[:chessboard][pos[0]][pos[1]]&0b1000, Chess::Game.ennemy(game[:chessboard][pos[0]][pos[1]]&0b1000)
      tmp, res = [
        [pos[0]+1,pos[1]+2],
        [pos[0]+1,pos[1]-2],
        [pos[0]-1,pos[1]+2],
        [pos[0]-1,pos[1]-2],
        [pos[0]+2,pos[1]+1],
        [pos[0]-2,pos[1]+1],
        [pos[0]+2,pos[1]-1],
        [pos[0]-2,pos[1]-1]
      ], []

      tmp.each do |move|
        res.push(move) unless (move[0] < 0 || move[0] > 7 || move[1] < 0 || move[1] > 7) || (!game[:chessboard][move[0]][move[1]].nil? && game[:chessboard][move[0]][move[1]]&0b1000 == color)
      end

      res

    end

  end

end

File: lib/game/knight/knight.c
/**
*
*  Knight-specific functions
*  @author Adrien Thébault <me@adrien-thebault.fr>
*
*/

/** Dependencies */

#include <stdbool.h>
#include <string.h>
#include "../game.h"

/**
*
*  Computes every possible move for a Knight at pos and puts result in possible_moves
*
*/

void ChessGame_Knight_PossibleMoves(game *g, unsigned char pos[2], unsigned char possible_moves[POSSIBLE_MOVES_SIZE][2][2], unsigned char *possible_moves_length) {

  signed char offset_line[8] = {1,1,-1,-1,2,-2,2,-2};
  signed char offset_col[8] = {2,-2,2,-2,1,1,-1,-1};

  signed char line, col, player = g->chessboard[pos[0]][pos[1]] & MASK_PLAYER;

  for(int i = 0; i<8; i++) {

    line = pos[0]+offset_line[i];
    col = pos[1]+offset_col[i];

    if(line > -1 && line < 8 && col > -1 && col < 8 && (g->chessboard[line][col] == -1 || (g->chessboard[line][col] & MASK_PLAYER) != player)) {

      memcpy(possible_moves[*possible_moves_length][0], pos, 2*sizeof(unsigned char));
      possible_moves[*possible_moves_length][1][0] = line;
      possible_moves[*possible_moves_length][1][1] = col;

      (*possible_moves_length)++;

    }

  }

}
